sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCoalesceBatches.scala

```diff
@@ -336,8 +338,22 @@ abstract class AbstractGpuCoalesceIterator(
         addBatch(batch)
       }
 
-      // there is a hard limit of 2^31 rows
-      while (numRows < Int.MaxValue && !hasOnDeck && iter.hasNext) {
+      def shouldFetchNextBatch: Boolean =
+        if (numRows < Int.MaxValue && !hasOnDeck && iter.hasNext) {
+          goal match {
+            case g: TargetSize =>
+              // avoid fetching if next batch is likely to exceed the specified row or size limits
+              // based on the stats of previous batches
+              val rowLimit = if (batchRowLimit > 0) batchRowLimit else Int.MaxValue
+              numRows + largestInputBatchSize <= rowLimit &&
```

Comment:
It looks like you switches size and rows here. The check has `largestInputBatchSize` paired with row checks, and `largetInputBatchRows` paired with bytes checks.